common: scripts: add prplmesh_utils.sh

tomereli opened the pull request:
Add initial prplmesh_utils.sh which can be used to start/stop/restart and check prplMesh status.
Currently, the script supports only DUMMY mode - but DWPAL mode can easilly be added by porting prplMesh-platform-init.sh into prplmesh_utils.sh.

The script brings up wlan0, wlan2 and eth1 as dummy interfaces and adds them to br-lan (which it also creates). Then it runs the framework local bus and transport (if building with MSGLIB), and starts the agent and controller (which is also a parameter - see help).

Usage - `sudo ./install/scripts/prplmesh_utils.sh <start/stop/restart/status>`

arnout commented on tools/scripts/prplMesh-platform-init-dummy.sh:
```diff
@@ -0,0 +1,100 @@
+    fi
+    ip link add wlan0 type dummy
+    ip link add wlan2 type dummy
+    ip link add eth1 type dummy
```
wlan0 and eth1 are interface names that might exist on the host. It would be better to replace these with names that are more likely to be unique.

Obviously, that means changing the agent config as well.

Suggested name: `pmt-ethN` for wired and `pmt-wlanN` for wireless. `pmt` stands for prplMeshTest. Note that the names can't be very long, there's a limit of 11 characters in some places.

tomereli replied:
I think we should do that, but preferably in a different PR since changing wlan interfaces names is risky and should be tested in RDKB / UGW.
Unless this is really happening to someone who has wlan0 and wlan2 or eth1 - lets put this one on hold for now.

arnout commented on common/beerocks/btl/CMakeLists.txt:
```diff
@@ -12,9 +12,11 @@ find_package(MapfTransport QUIET)
 
 if(MapfTransport_FOUND)
 message(STATUS "${BoldGreen}Using MultiAP Framework for BTL transport${ColourReset}")
+set(BTL_TYPE "LOCAL_BUS" CACHE STRING "" FORCE)
```
This should be marked as INTERNAL as well. Note that INTERNAL implies FORCE.
```suggestion
set(BTL_TYPE "LOCAL_BUS" CACHE STRING "" INTERNAL)
```

Also, even for internal cache variables, it's good practice to provide a non-empty docstring.

tomereli replied:
I couldn't find out what the docstring is and what I should set it to, can you enlighten me?

arnout commented on common/beerocks/btl/CMakeLists.txt:
```diff
@@ -12,9 +12,11 @@ find_package(MapfTransport QUIET)
 set(btl_sources btl.cpp btl_local_bus.cpp)
 else()
 message(STATUS "${BoldYellow}Failed to find MultiAP Framework transport, fall back to BTL over UDS - supporting single node only!${ColourReset}")
+set(BTL_TYPE "UDS" CACHE STRING "" FORCE)
```
```suggestion
set(BTL_TYPE "UDS" CACHE STRING "" FORCE INTERNAL)
```

arnout commented on common/beerocks/bwl/CMakeLists.txt:
```diff
@@ -17,7 +17,7 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++11")
 # UGW
 if (TARGET_PLATFORM STREQUAL "ugw")
-    set(BWL_TYPE "DWPAL")
+   set(BWL_TYPE "DWPAL")
```
Incorrect whitespace change.

arnout replied:
Just to be clear: my suggestion from below is to change this like this:
```suggestion
    set(BWL_TYPE_DEFAULT "DWPAL")
```

Commit:
btl: build: add CMAKE BTL_TYPE variable

Add BTL_TYPE as cached variable so it can be used in init script for
either starting local bus and ieee1905_transport or not, based on
framework's MSGLIB option.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
bwl: build: make BWL_TYPE globally accessible

Make BWL_TYPE globally accessible for other CMakeLists.txt so that it
can be used in init script for selecting which platform init function to
call. Currently only DUMMY mode is supported.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on common/beerocks/bwl/CMakeLists.txt:
```diff
@@ -79,7 +81,8 @@ elseif(BWL_TYPE STREQUAL "WAV")
 elseif(BWL_TYPE STREQUAL "DWPAL")
-
+    
+    set(BWL_TYPE "DWPAL" CACHE STRING "" FORCE)
```
Setting BWL_TYPE twice, first as normal variable and then as cache variable, is just confusing.

I think it's better to set local variable BWL_TYPE_DEFAULT in the platform condition, and after the condition add one and only one
```
set(BWL_TYPE ${BWL_TYPE_DEFAULT} CACHE STRING "Which BWL backend to use")
```

tomereli replied:
done

Commit:
scripts: add initial prplmesh_utils.sh

Add initial prplmesh_utils.sh which can be used to start/stop/restart
and check prplMesh status.
Currently the script supports only DUMMY mode - but DWPAL mode can be
easilly added by porting prplMesh-platform-init.sh into
prplmesh_utils.sh.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -0,0 +1,159 @@
+        err "$0: This script must be run as root"
+        exit 1
+    fi
+    [[ "$BWL_TYPE" == "DUMMY" ]] && platform_deinit_dummy
```
Just for my information: why use `[[` instead of `[`?

tomereli replied:
just more features provided by `bash` - see [here](https://stackoverflow.com/questions/3427872/whats-the-difference-between-and-in-bash)

I still quote my strings though :)

arnout replied:
For the record: the second answer of [another stackoverflow question](https://stackoverflow.com/questions/669452/is-double-square-brackets-preferable-over-single-square-brackets-in-ba) advises against using `[[`:
> Recommendation: always use [].
>
> There are POSIX equivalents for every [[ ]] construct I've seen.
>
> If you use [[ ]] you:
>
> * lose portability
> * force the reader to learn the intricacies of another bash extension. [ is just a regular command with a weird name, no special semantics are involved.

I'm fine either way though. We should just try to be consistent.

tomereli replied:
convinced me, I'll change to `[`

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -0,0 +1,159 @@
+
+    if [ $? != 0 ] ; then err "Failed parsing options." >&2 ; usage; exit 1 ; fi
+
+    #echo "$OPTS"
```
Remove this debug statement.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -0,0 +1,159 @@
+}
+
+VERBOSE=false
+HELP=false
```
This variable is unused.

arnout commented on common/beerocks/scripts/CMakeLists.txt:
```diff
@@ -1,4 +1,10 @@
+configure_file(
+        "${CMAKE_CURRENT_SOURCE_DIR}/prplmesh_utils.sh.in"
+        "${CMAKE_CURRENT_BINARY_DIR}/prplmesh_utils.sh"
+        )
```
Since this is a shell script which may use `${VAR}` syntax, add `@ONLY`:
```suggestion
        @ONLY 
        )
```

tomereli replied:
ok

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -0,0 +1,159 @@
+#!/bin/bash
+
+BWL_TYPE=@BWL_TYPE@
+BTL_TYPE=@BTL_TYPE@
```
There is no need to declare variables for these. Just use `@BWL_TYPE@` directly below.

tomereli replied:
2 reasons I prefer to still declare them:
1. more readable than `[[ "DUMMY" == "DUMMY" ]] && ...`
2. future options to the script to override the BWL type on deployed system (when we won't build natively)

arnout replied:
> 2 reasons I prefer to still declare them:
> 
>     1. more readable than `[[ "DUMMY" == "DUMMY" ]] && ...`

I disagree. With the variable, it is not immediately apparent what will be done. With the "DUMMY" == "DUMMY" it's immediately clear. You may wonder a bit why something like that would be written, but it's not a big leap for someone familiar with autotools or CMake configure files.


>     2. future options to the script to override the BWL type on deployed system (when we won't build natively)

There are three cases to consider.
1. Development builds on Ubuntu.
2. Development builds on platform.
3. Releases on platform.

The third case will always use the platform's process management, never this script. So I assume you're talking about the second case.

Then, however, whenever you make a build, the script will be regenerated and reinstalled. I admit, due to the CMAKE_MULTIAP_OUTPUT_DIRECTORY hack, it's not actually going to be copied to the target. But if you would instead go through the install directory, the script would always be correct with the bwl/btl that is actually installed on the target.

Yes, it would allow you to do a debug build without BTL and run it on target with the local bus running, but I think that that is so much of a hack that the script doesn't need to support it.

In addition, I definitely *don't* like adding complexity to the code base for some future unproven use case.

Anyway, not a big deal if it stays like it is - it's not adding that much complexity.

simonbaren replied:
If you want it for future use of debug, 
Either do `BWL_TYPE=${1-@BWL_TYPE@}` 
or
```
if test -z "$BWL_TYPE" 
then
      BWL_TYPE=@BWL_TYPE@
fi
```
Or better - get it in the opts

tomereli replied:
I took @arnout 's advice eventually.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -0,0 +1,159 @@
+        err "$0: This command must be run as root"
+        exit 1
+    fi
+    [[ "$BWL_TYPE" == "DUMMY" ]] && platform_init_dummy
```
```suggestion
    [[ "@BWL_TYPE@" == "DUMMY" ]] && platform_init_dummy
```

tomereli replied:
As I said, this will be less read for users who look at the output script

arnout reviewed (changes requested):
Great feature!

We'll also need an update of the README.md, eventually.

Commit:
doc: README.md: update with running instructions

Update top level README.md with section for how to run prplMesh, where
logs are located, and how to test that the prplMesh is operational.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented:
@tomereli if you did a rebase-force-push after the review was done, it doesn't make sense to follow up with fixup commits since the "Show changes since your last review" anyway doesn't work anymore. It's better to then continue doing rebase-force-pushes, until someone starts reviewing again.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -0,0 +1,157 @@
+BTL_TYPE=@BTL_TYPE@
+
+dbg() {
+    [ "$VERBOSE" == "true" ] && echo "$@"
```
It's not logical that you use `[` here and `[[` below :-)

arnout commented on README.md:
````diff
@@ -34,19 +32,43 @@ repo forall -p -c 'git checkout $REPO_RREV'
+Once built, prplMesh controller, agent and framework can be started using `prplmesh_utils.sh`:
+
+```bash
+cd ~/prplMesh/build/install/scripts
````
Above, we assume the current directory is just above prplMesh (i.e. the root of the manifest checkout), so the path should be relative to that. We definitely can't assume that it's checked out in `~/prplMesh`.
```suggestion
cd build/install/scripts
```

arnout commented on README.md:
````diff
@@ -34,19 +32,43 @@ repo forall -p -c 'git checkout $REPO_RREV'
+sudo ./prplmesh_utils.sh start
+```
+
+### Log files locations
````
That's a good addition!

arnout commented on common/beerocks/bwl/CMakeLists.txt:
```diff
@@ -47,6 +51,7 @@ message("${BoldWhite}Preparing ${BoldGreen}${PROJECT_NAME}${BoldWhite} ${BWL_TYP
 
 if (BWL_TYPE STREQUAL "WAV_FAPI")
 
+    set(BWL_TYPE "WAV_FAPI" CACHE STRING "" FORCE)
```
This is no longer needed.

arnout commented on common/beerocks/bwl/CMakeLists.txt:
```diff
@@ -56,6 +61,7 @@ if (BWL_TYPE STREQUAL "WAV_FAPI")
 
 elseif(BWL_TYPE STREQUAL "WAV")
 
+    set(BWL_TYPE "WAV" CACHE STRING "" FORCE)
```
Not needed.

arnout commented on common/beerocks/bwl/CMakeLists.txt:
```diff
@@ -79,7 +85,8 @@ elseif(BWL_TYPE STREQUAL "WAV")
 elseif(BWL_TYPE STREQUAL "DWPAL")
-
+    
+    set(BWL_TYPE "DWPAL" CACHE STRING "" FORCE)
```
Not needed.

arnout commented on common/beerocks/bwl/CMakeLists.txt:
```diff
@@ -90,6 +97,7 @@ elseif(BWL_TYPE STREQUAL "DWPAL")
 
 elseif(BWL_TYPE STREQUAL "DUMMY")
 
+    set(BWL_TYPE "DUMMY" CACHE STRING "" FORCE)
```
Not needed.

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -0,0 +1,157 @@
+
+prplmesh_agent_start() {
+    echo "prplmesh_agent_start - start beerocks_agent process..."
+    @CMAKE_INSTALL_PREFIX@/bin/beerocks_agent &
```
It may have been a fluke on my side, but at some point I had to cd to a specific directory (the install directory IIRC) for beerocks_agent to find its config file...

tomereli replied:
It was like that in the beginning but we fixed that

tomereli commented:
> @tomereli if you did a rebase-force-push after the review was done, it doesn't make sense to follow up with fixup commits since the "Show changes since your last review" anyway doesn't work anymore. It's better to then continue doing rebase-force-pushes, until someone starts reviewing again.

Still getting used to the workflow, will try not to do force-pushes after the review has started.

arnout commented on common/beerocks/btl/CMakeLists.txt:
```diff
@@ -12,11 +12,11 @@ find_package(MapfTransport QUIET)
 if(MapfTransport_FOUND)
 message(STATUS "${BoldGreen}Using MultiAP Framework for BTL transport${ColourReset}")
-set(BTL_TYPE "LOCAL_BUS" CACHE STRING "" INTERNAL)
+set(BTL_TYPE "LOCAL_BUS" PARENT_SCOPE)
```
So that *does* work in the end!

What is the effect exactly? Visibility in the btl/CMakeLists.txt, or up to the common/CMakeLists.txt?

tomereli replied:
>"If the PARENT_SCOPE option is given the variable will be set in the scope above the current scope."

So in this case, it will be visible in btl/CMakeLists.txt and any subdirectory (this is trying for any SET) but also for the upper directory, which is indeed common/CMakeLists.txt

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -1,10 +1,7 @@
-BTL_TYPE=@BTL_TYPE@
-
 dbg() {
-    [ "$VERBOSE" == "true" ] && echo "$@"
```
Hah, I never even noticed this one...

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -82,10 +79,10 @@ start_function() {
-    [[ "$PRPLMESH_MODE" == "CA" || "$PRPLMESH_MODE" == "A" ]] && prplmesh_agent_start
+    [ "@BWL_TYPE@" = "DUMMY" ] && platform_init_dummy
+    [ "@BTL_TYPE@" = "LOCAL_BUS" ] && prplmesh_framework_init
+    [ "$PRPLMESH_MODE" = "CA" -o "$PRPLMESH_MODE" = "C" ] && prplmesh_controller_start
```
Note that shellcheck advises in general to use:
```
[ "$PRPLMESH_MODE" = "CA" ] || [ "$PRPLMESH_MODE" = "C" ]
```
but TBH I don't understand why...

arnout commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -1,10 +1,7 @@
 #!/bin/bash
```
Since all the bashisms in this file have been removed, we could change this to `/bin/sh`.

However, I'm in favour of standardising to use `/bin/bash` even in POSIX-compliant shell scripts. It's all too easy for a bashism to sneak in, and it doesn't become apparent until someone runs that specific code path in a shell that doesn't support that particular bashism.

Only if we would add scripts that are really intended to be used on released products, they should use POSIX shell. However, I don't think it should be necessary to have any prplMesh scripts in released products.
